osu.Game.Rulesets.Osu/Difficulty/MathUtil/HitProbabilities.cs

```diff
@@ -0,0 +1,161 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+
+using osu.Game.Rulesets.Osu.Difficulty.Preprocessing;
+
+namespace osu.Game.Rulesets.Osu.Difficulty.MathUtil
+{
+    class HitProbabilities
+    {
+        public static int cacheHit = 0;
+        public static int cacheMiss = 0;
+        List<MapSectionCache> sections = new List<MapSectionCache>();
+        public HitProbabilities(List<OsuMovement> movements, double cheeseLevel, int difficultyCount = 20)
+        {
+            for (int i = 0; i < difficultyCount; ++i)
+            {
+                int start = movements.Count * i / difficultyCount;
+                int end = movements.Count * (i + 1) / difficultyCount - 1;
+                sections.Add(new MapSectionCache(movements.GetRange(start, end - start + 1), cheeseLevel));
+            }
+        }
+
+        public int Count(int start, int sectionCount)
+        {
+            int count = 0;
+            for (int i = start; i != start + sectionCount; i++)
+            {
+                count += sections[i].Movements.Count;
+            }
+
+            return count;
+        }
+        public bool IsNull(int sectionCount)
```

Comment:
`IsEmpty` would be better imo, `IsNull` sounds like its checking for `null`

Reply:
agree with everything